src/components/RecordingDialog.tsx

```diff
@@ -189,4 +191,75 @@ const RecordingDialog = ({
   );
 };
 
+interface RecordingOptions {
+  onDone: (data: XYZData) => void;
+  onProgress: (percentage: number) => void;
+}
+
+interface InProgressRecording extends RecordingOptions {
+  data: XYZData;
+  startTimeMillis: number;
+}
+
+interface RecordingDataSource {
+  startRecording(options: RecordingOptions): void;
+  cancelRecording(): void;
+}
+
+const useRecordingDataSource = (): RecordingDataSource => {
+  const ref = useRef<InProgressRecording | undefined>();
+  const connectActions = useConnectActions();
+  useEffect(() => {
+    const listener = (e: AccelerometerDataEvent) => {
+      const {
+        data: { x, y, z },
+      } = e;
+      if (ref.current) {
+        ref.current.data.x.push(x / 1000);
+        ref.current.data.y.push(y / 1000);
+        ref.current.data.z.push(z / 1000);
+        const percentage =
+          ((Date.now() - ref.current.startTimeMillis) / mlSettings.duration) *
+          100;
+        ref.current.onProgress(percentage);
+      }
+    };
+    connectActions.addAccelerometerListener(listener);
+    return () => {
+      connectActions.removeAccelerometerListener(listener);
+    };
+  }, [connectActions]);
+
+  return useMemo(
+    () => ({
+      timeout: undefined as ReturnType<typeof setTimeout> | undefined,
+
+      startRecording(options: RecordingOptions) {
+        this.timeout = setTimeout(() => {
+          if (ref.current) {
+            const sampleCount = ref.current.data.x.length;
+            if (sampleCount < mlSettings.minSamples) {
+              // TODO: this is an error case without UX
```

Comment:
On main we show an toast after/as the dialog goes away with the string:

  "alert.recording.disconnectedDuringRecording": "micro:bit disconnected during recording"

Although there could be other causes this seems reasonable.